[llvm-lit][test] Precommit tests for lit's built-in cat command

connieyzhu opened the pull request:
This patch creates preliminary checks for lit's built-in cat command. These tests verify the cat and cat -v functionality, and display failures with using cat -e which will be implemented in a later patch. This is a prequisite for https://github.com/llvm/llvm-project/pull/102061, and is dependent on https://github.com/llvm/llvm-project/pull/102366.

connieyzhu committed:
[llvm-lit][test] Adding tests for lit's built-in cat command

This patch creates preliminary checks for lit's built-in cat command.
These tests verify the cat and cat -v functionality, and display
failures with using cat -e which will be implemented in a later patch.

connieyzhu commented:
CC: @ilovepi @petrhosek

ilovepi commented on llvm/utils/lit/tests/shtest-cat.py:
```diff
@@ -0,0 +1,32 @@
+# Check the cat command
```
```suggestion
## Test the cat command
```
Check can sometimes be easy to miss w/ `CHECK` directives, and we often use `##` to indicate that a line is commentary, and not intended to be part of the lit/FileCheck directives.

connieyzhu replied:
done

ilovepi commented on llvm/utils/lit/tests/shtest-cat.py:
```diff
@@ -0,0 +1,32 @@
+# CHECK: # executed command: FileCheck {{.*}}
+# CHECK: # error: command failed with exit status: {{.*}}
+
+# CHECK: PASS: shtest-cat :: cat-newline.txt ({{[^)]*}})
+# CHECK: cat {{.+}}/newline | FileCheck {{.*}}
+# CHECK: # executed command: cat {{.*}}
+# CHECK: # executed command: FileCheck {{.*}}
```
Perhaps also for `cat -v`?

connieyzhu replied:
done

ilovepi commented on llvm/utils/lit/tests/Inputs/shtest-cat/cat-newline.txt:
```diff
@@ -0,0 +1,12 @@
+# RUN: cat %S/txt-files/newline | FileCheck %s
```
Maybe direct this to a file w/ `> %t`, and then you can assert some things about the output? I was thinking verifying the length and other properties, but you could probably just diff it against the original file.

ilovepi commented on llvm/utils/lit/tests/Inputs/shtest-cat/cat-e-newline.txt:
```diff
@@ -0,0 +1,12 @@
+# RUN: cat -e %S/txt-files/newline | FileCheck %s
```
Does this need to be its own file? I understand you're separating out the `-e` behaviors, but maybe we should either group all of the tests based on their flags (e.g. one test file for `cat`, `cat -v`, and `cat -e`), or group them on per-file behavior (e.g. one test for `newlines`, one for `allchars`, etc.). I feel like thats likely to reduce duplication.

connieyzhu replied:
done

ilovepi commented:
I'd suggest changing the title to `[llvm-lit][test] Precommit tests for lit's built-in cat command`.

connieyzhu committed:
[llvm-lit] Combined test files to separate by option for cat

Tests for the same cat options were initially in different files, but
now these tests are condensed into 3 files that contain all tests for
each cat option, which are cat, cat -v, and cat -e. Also made some minor
syntax changes and added clarifying comments.

ilovepi commented on llvm/utils/lit/tests/Inputs/shtest-cat/cat-e.txt:
```diff
@@ -0,0 +1,33 @@
+
+# RUN: cat -e %S/txt-files/allchars | FileCheck %s --check-prefix=ALLCHARS
+# RUN: cat -e %S/txt-files/newline | FileCheck %s --check-prefix=NEWLINE
+
```
Can you add a comment here noting that allchars is for testing metacharacter output w/ `-e`?
```suggestion
## Test output of -e for all ASCII characters including  ^ and M- notation

```

connieyzhu replied:
done

ilovepi commented on llvm/utils/lit/tests/Inputs/shtest-cat/cat-e.txt:
```diff
@@ -0,0 +1,33 @@
+# RUN: cat -e %S/txt-files/allchars | FileCheck %s --check-prefix=ALLCHARS
+# RUN: cat -e %S/txt-files/newline | FileCheck %s --check-prefix=NEWLINE
+
+# ALLCHARS: ^@^A^B^C^D^E^F^G^H      $
```
```suggestion
# ALLCHARS: ^@^A^B^C^D^E^F^G^H
# ALLCHARS-SAME: $
```
May want to exclude chars that we can't output/see in the FileCheck line.`{{.*}}$` could also work.

My concern is those just look like whitespace, but are probably something else.

connieyzhu replied:
done

ilovepi commented on llvm/utils/lit/tests/Inputs/shtest-cat/cat-v.txt:
```diff
@@ -0,0 +1,49 @@
+#     fclose(fptr);
+#     return 0;
+# }
+
```
nit: extra newline

connieyzhu replied:
done

ilovepi commented on llvm/utils/lit/tests/Inputs/shtest-cat/cat-v.txt:
```diff
@@ -0,0 +1,49 @@
+
+# RUN: cat -v %S/txt-files/allchars | FileCheck %s --check-prefix=ALLCHARS
+# RUN: cat -v %S/txt-files/newline | FileCheck %s --check-prefix=NEWLINE
+
```
Comments here, similar to the one for `-e` would help

connieyzhu replied:
done

ilovepi commented on llvm/utils/lit/tests/Inputs/shtest-cat/cat.txt:
```diff
@@ -0,0 +1,14 @@
+## Tests
```
Please either describe what you're testing, or remove the top comment + newline.

connieyzhu replied:
done

ilovepi commented on llvm/utils/lit/tests/Inputs/shtest-cat/cat-v.txt:
```diff
@@ -0,0 +1,49 @@
+# C Program to generate all ASCII characters in allchars.txt
```
allchars is marked as a binary file in the PR. Can we either add some leading text or a file extension to make it clear to git/Github that it isn't a binary file? I know you had issues w/ `.txt` extensions getting identified as tests, but perhaps another type of extension would work? or maybe just some leading string before the strange characters?

connieyzhu replied:
I believe the special characters themselves make the file binary. Even when the files used to have the .txt extension, it would still be marked as a binary file. I tried adding some leading strings but it didn't change anything.

ilovepi replied:
Interesting. If neither approach works, then there isn't much to do, I suppose. That's unfortunate, but isn't a blocker.

ilovepi commented on llvm/utils/lit/tests/Inputs/shtest-cat/cat-v.txt:
```diff
@@ -0,0 +1,49 @@
+# }
+
+
+## Tests to check cat -e output for ASCII characters and newlines
```
```suggestion
## Tests to check cat -v output for ASCII characters and newlines
```

connieyzhu replied:
done

ilovepi commented on llvm/utils/lit/tests/Inputs/shtest-cat/lit.cfg:
```diff
@@ -0,0 +1,8 @@
+config.test_format = lit.formats.ShTest()
+config.test_source_root = None
+config.test_exec_root = None
+config.substitutions.append(("%{python}", '"%s"' % (sys.executable)))
```
Are you using python? If not, please remove.

connieyzhu replied:
done

ilovepi commented on llvm/utils/lit/tests/Inputs/shtest-cat/cat-v.txt:
```diff
@@ -0,0 +1,49 @@
+# C Program to generate all ASCII characters in allchars.txt
```
```suggestion
## C Program to generate all ASCII characters in allchars
```
The file is `allchars`. If you change the extension, please match it.

connieyzhu replied:
done

ilovepi reviewed (commented):
This is basically LGTM, but I wonder if we need more tests for `cat`? Maybe one that diffs the original file to a temp file?

Maybe something like:
```
RUN: cat %s/txt-files/allchars > %t
RUN: diff %s/txt-files/allchars %t
```

connieyzhu committed:
[llvm-lit] Added more clarifying comments + minor syntax changes

More comments describing the purpose of the tests were added for
clarification purposes. Some small syntax changes were also made for
better readability.

connieyzhu committed:
[llvm-lit] Added allchars test for cat

This patch adds another test to check the cat functionality with the
allchars text file that includes special/meta characters.

connieyzhu commented on llvm/utils/lit/tests/Inputs/shtest-cat/cat.txt:
Just realized that with the FileCheck crash (https://github.com/llvm/llvm-project/issues/101582), I can't check that there is no output for diff (which would be the case if the two files were the same). Is there another way to check output without using FileCheck?

ilovepi replied:
diff returns an error code when files are different. You shouldn't need to use a check line at all, since the command itself should fail.

Here's an example from LLD's tests, but other places do somthing similar. https://github.com/llvm/llvm-project/blob/72a514f42722ad6f6e6baee8e2d75150f9369a66/lld/test/ELF/stdout.s#L10

ilovepi commented on llvm/utils/lit/tests/shtest-cat.py:
```diff
@@ -0,0 +1,31 @@
+## Test the cat command
+
+# RUN: not %{lit} -a -v %{inputs}/shtest-cat \
+# RUN: | FileCheck -match-full-lines %s --dump-input-context=1000
```
Why is `--dump-input-context=1000` needed?

connieyzhu replied:
Oh, that was just to get more context as I was debugging. I can remove that in my next patch

connieyzhu committed:
[llvm-lit] Changed CHECK directives for more check precision

Made changes to CHECK directives to check output more precisely.

connieyzhu commented on llvm/utils/lit/tests/Inputs/shtest-cat/cat-e-allchars.txt:
```diff
@@ -0,0 +1,19 @@
+# RUN: cat -e %S/txt-files/allchars | FileCheck %s
```
done

connieyzhu commented on llvm/utils/lit/tests/Inputs/shtest-cat/txt-files/newline:
```diff
@@ -0,0 +1,10 @@
+
+
+
+line
```
done

connieyzhu commented on llvm/utils/lit/tests/shtest-cat.py:
```diff
@@ -0,0 +1,27 @@
+
+# CHECK: PASS: shtest-cat :: cat-v.txt ({{[^)]*}})
+# CHECK: cat -v {{.+}}/allchars | FileCheck {{.*}}
+# CHECK: # executed command: cat -v {{.*}}
```
done

connieyzhu committed:
[llvm-lit] Fixed darker formatting issue

jh7370 commented on llvm/utils/lit/tests/shtest-cat.py:
```diff
@@ -0,0 +1,31 @@
+# CHECK: cat {{.+}}/allchars > {{.+}}
+# CHECK-NEXT: # executed command: cat {{.+}}/allchars
+# CHECK: diff {{.+}}/allchars {{.+}}
+# CHECK-NEXT: # executed command: diff {{.+}}/allchars {{.+}}
```
Missing new line at EOF.

jh7370 commented on llvm/utils/lit/tests/Inputs/shtest-cat/cat.txt:
```diff
@@ -0,0 +1,16 @@
+## Tests to check cat output
+
+# RUN: cat %S/txt-files/newline | FileCheck %s --check-prefix=NEWLINE
```
Can we simply use `diff` here instead of FileCheck, like you do below?

jh7370 reviewed (commented):
How many tests use the `-e` / `-v` functionality of `cat`?

connieyzhu commented on llvm/utils/lit/tests/Inputs/shtest-cat/txt-files/newline:
```diff
@@ -0,0 +1,10 @@
+
+
+
+last
```
It is not, but I believe it does test a special case where a file does not end with a newline. Do you think this is worthwhile to test, or should a newline be added?

connieyzhu committed:
[llvm-lit][test] Added cat -e tests to existing cat testing file

This patch removes cat test files created in previous patches in this PR
and adds -e tests to valid-shell.txt, which already has existing cat
tests.

connieyzhu commented:
I realized I somehow grepped incorrectly, and there actually already exists tests for `cat`. In my most recent patch, I've removed the test files I made and added a new test for `-e` in the existing file. Sorry for the inconvenience, I appreciate your patience.

> As above, consider adding `--implicit-check-not={{.}}`, `--strict-whitespace`, and `--match-full-lines` to the FileCheck commands.

@jh7370 I've added these options to the new `-e` test I added to `valid-shell.txt`. The other existing tests in this file don't seem to have these options set for FileCheck, and also have grammar/punctuation issues that you mentioned before. Do you think I make these changes to the existing tests too?

jh7370 commented:
> I realized I somehow grepped incorrectly, and there actually already exists tests for `cat`. In my most recent patch, I've removed the test files I made and added a new test for `-e` in the existing file. Sorry for the inconvenience, I appreciate your patience.

No problem.

> > As above, consider adding `--implicit-check-not={{.}}`, `--strict-whitespace`, and `--match-full-lines` to the FileCheck commands.
> 
> @jh7370 I've added these options to the new `-e` test I added to `valid-shell.txt`. The other existing tests in this file don't seem to have these options set for FileCheck, and also have grammar/punctuation issues that you mentioned before. Do you think I should make these changes to the existing tests too?

I think the typical LLVM process would be to improve the existing tests first in a PR prior to this one, then base this PR on top of it. You could use a tool like spr to stack your PRs, to aid with this, or just keep them separate then rebase this PR once you land the first one.

jh7370 commented on llvm/utils/lit/tests/Inputs/shtest-shell/valid-shell.txt:
```diff
@@ -169,3 +169,23 @@
+#
+# RUN: cat -e %S/cat_nonprinting.bin \
+# RUN: | FileCheck --check-prefix=NPE-CAT-OUTPUT --implicit-check-not={{.}} --match-full-lines --strict-whitespace %s
+# NPE-CAT-OUTPUT:^@^A^B^C^D^E^F^G	^H$
```
Nit, for readability, I might add some spacing as in the suggestion, so that the colons line up.
```suggestion
#      NPE-CAT-OUTPUT:^@^A^B^C^D^E^F^G	^H$
```

jh7370 reviewed (commented):
Picking up on your earlier comment, I think showing the behaviour of `cat` with a file without a trailing new line is indeed an interesting behaviour to test. It will require a new test input.

Looking more generally at the existing test file, if I'm not mistaken, you should be able to remove the `NPLONG-CAT-OUTPUT` checks, since they're identical to `NP-CAT-OUTPUT`. I'd also consider splitting all these cat tests into a separate file (like you had before) for ease of identification in the future. You can keep all the cat tests in the same test file, but I'd suggest it not be mixed with the other test behaviour. I'd do all that in the same prerequisite PR that I mentioned, to avoid conflating new tests with existing tests. I'd even consider two prerequisite PRs: the first to move the tests, and the second to improve them as suggested. This would make reviewing the second of these easier. Up to you though.
